### PR TITLE
KN-11541 Using memcached for outputcache and localization cache

### PR DESCRIPTION
**Pull request by AlexPhilippov**
Moved to memcached cache provider

### CODE REVIEW

File: Core/LocalizationManager.cs [hunk 1]
```diff
@@ -156,16 +156,63 @@ public IEnumerable<Object> GetScriptResources(CultureInfo culture)
...
 
 			return lst;
 		}
 
+		private static string GetBundleName(CultureInfo culture, int bundleIndex)
+		{
+			return string.Format("{0}_bundle{1}", culture.Name, bundleIndex);
+		}
+
+		private int GetBundleIndex(string Hash)
```
Comment (polyakova) [inline Core/LocalizationManager.cs]: Hash -> hash

File: Core/LocalizationManager.cs [hunk 2]
```diff
@@ -156,16 +156,63 @@ public IEnumerable<Object> GetScriptResources(CultureInfo culture)
 
 		public IEnumerable<ILocalizedObject> GetAll(CultureInfo culture)
 		{
-			var lst = LocalizationCache.Get<IEnumerable<ILocalizedObject>>(culture.Name);
+			IEnumerable<ILocalizedObject> lst = null;
+			if (_initialBundlesCount > 0)
+			{
+				List<ILocalizedObject>[] bundles = new List<ILocalizedObject>[_initialBundlesCount];
```
Comment (polyakova) [inline Core/LocalizationManager.cs]: our internal code convention is to use "var" instead of full type

File: Core/LocalizationManager.cs [hunk 3]
```diff
@@ -156,16 +156,63 @@ public IEnumerable<Object> GetScriptResources(CultureInfo culture)
...
+		private int GetBundleIndex(string Hash)
+		{
+			int index = Hash.GetHashCode();
+			if (index < 0)
+				index = -index;
+			index %= _initialBundlesCount;
+			return index;
+		}
+
+		public IEnumerable<ILocalizedObject> GetCachedListForHash(CultureInfo culture, string Hash)
```
Comment (polyakova) [inline Core/LocalizationManager.cs]: Hash -> hash

File: Core/LocalizationManager.cs [hunk 4]
```diff
@@ -156,16 +156,63 @@ public IEnumerable<Object> GetScriptResources(CultureInfo culture)
 
 		public IEnumerable<ILocalizedObject> GetAll(CultureInfo culture)
 		{
-			var lst = LocalizationCache.Get<IEnumerable<ILocalizedObject>>(culture.Name);
+			IEnumerable<ILocalizedObject> lst = null;
```
Comment (polyakova) [inline Core/LocalizationManager.cs]: name it "result"

File: Core/LocalizationManager.cs [hunk 5]
```diff
@@ -156,16 +156,63 @@ public IEnumerable<Object> GetScriptResources(CultureInfo culture)
...
+				List<ILocalizedObject>[] bundles = new List<ILocalizedObject>[_initialBundlesCount];
+				for (int i = 0; i < _initialBundlesCount; i++)
+				{
+					bundles[i] = LocalizationCache.Get<List<ILocalizedObject>>(GetBundleName(culture, i));
+					if (bundles[i] == null)
+						break;
+				}
+				if (!bundles.Any(l => l == null))
+				{
+					var lstCombined = new List<ILocalizedObject>();
```
Comment (polyakova) [inline Core/LocalizationManager.cs]: You don't extra var "lstCombined", use "result" list

File: Core/LocalizationManager.cs [hunk 6]
```diff
@@ -156,16 +156,63 @@ public IEnumerable<Object> GetScriptResources(CultureInfo culture)
...
 		{
-			var lst = LocalizationCache.Get<IEnumerable<ILocalizedObject>>(culture.Name);
+			IEnumerable<ILocalizedObject> lst = null;
+			if (_initialBundlesCount > 0)
+			{
+				List<ILocalizedObject>[] bundles = new List<ILocalizedObject>[_initialBundlesCount];
+				for (int i = 0; i < _initialBundlesCount; i++)
+				{
+					bundles[i] = LocalizationCache.Get<List<ILocalizedObject>>(GetBundleName(culture, i));
+					if (bundles[i] == null)
```
Comment (polyakova) [inline Core/LocalizationManager.cs]: when it will be null?

File: Core/LocalizationManager.cs (hunk 3 above)
Comment (polyakova) [inline Core/LocalizationManager.cs]: It better to pass ILocalizedObject and call method GetBundle(CultureInfo culture, ILocalizedObject  object)

File: Core/LocalizationManager.cs [hunk 7]
```diff
@@ -178,58 +225,70 @@ public IEnumerable<CultureInfo> GetCultures()
...
 
-		public void Delete(params ILocalizedObject[] list)
+		public void Delete(CultureInfo culture, params ILocalizedObject[] list)
 		{
 			Repository.Delete(list);
-			LocalizationCache.Clear();
+			RemoveFromCache(culture, list);
+		}
+
+		private void RemoveFromCache(CultureInfo culture, ILocalizedObject[] list)
```
Comment (polyakova) [inline Core/LocalizationManager.cs]: call it Remove

File: Core/LocalizationManager.cs [hunk 8]
```diff
@@ -156,16 +156,63 @@ public IEnumerable<Object> GetScriptResources(CultureInfo culture)
...
+			}
 			if (lst == null || lst.Count() == 0)
 			{
 				lst = Repository.GetAll(culture).ToList();
-				LocalizationCache.Insert(culture.Name, lst);
+				if (_initialBundlesCount <= 0)
+					_initialBundlesCount = lst.Count() * 1024 / (1024 * 1024) * 4;
+				List<ILocalizedObject>[] bundles = new List<ILocalizedObject>[_initialBundlesCount];
+				for (int i = 0; i < _initialBundlesCount; i++)
+					bundles[i] = new List<ILocalizedObject>();
```
Comment (polyakova) [inline Core/LocalizationManager.cs]: add empty line between two logic blocks.

for ()
{}

if ()
{}

etc

File: Core/LocalizationManager.cs (hunk 6 above)
Comment (AlexPhilippov) [inline Core/LocalizationManager.cs]: If someone resets cache simultaneously

File: Core/LocalizationManager.cs (hunk 5 above)
Comment (AlexPhilippov) [inline Core/LocalizationManager.cs]: result is IEnumerable<>, I need List to add members

File: Core/LocalizationManager.cs (hunk 3 above)
Comment (AlexPhilippov) [inline Core/LocalizationManager.cs]: This method is used to get LocalizedObject by hash, so we doesn't have object itself to call it.

File: Core/LocalizationManager.cs (hunk 7 above)
Comment (AlexPhilippov) [inline Core/LocalizationManager.cs]: There is already method called 'Delete', I guess just 'Remove' will be not enough. This class is responsible for all localization processing, so it's better to specify that we're clearing cache.

**Pull request merged**